core/drivers/xiphera_trng.c

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Vaisala Oyj
+ */
+
+#include <assert.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+
+#define CONTROL_REG                     0x00000000
+#define STATUS_REG                      0x00000004
+#define RAND_REG                        0x00000000
+
+#define HOST_TO_TRNG_RESET              0x00000001
+#define HOST_TO_TRNG_RELEASE_RESET      0x00000002
+#define HOST_TO_TRNG_ENABLE             0x80000000
+#define HOST_TO_TRNG_ZEROIZE            0x80000004
+#define HOST_TO_TRNG_ACK_ZEROIZE        0x80000008
+#define HOST_TO_TRNG_READ               0x8000000F
+
+/* trng statuses */
+#define TRNG_ACK_RESET                  0x000000AC
+#define TRNG_SUCCESSFUL_STARTUP         0x00000057
+#define TRNG_FAILED_STARTUP             0x000000FA
+#define TRNG_NEW_RAND_AVAILABLE         0x000000ED
+
+static unsigned int trng_lock = SPINLOCK_UNLOCK;
+
+static vaddr_t xiphera_trng_base;
+
+static uint32_t xiphera_trng_read32(void)
+{
+	uint32_t value = 0;
+	uint32_t exceptions = 0;
+	uint32_t status = 0;
+
+	exceptions = cpu_spin_lock_xsave(&trng_lock);
+
+	while (true) {
+		/* Wait until we have value available */
+		status = io_read32(xiphera_trng_base + STATUS_REG);
+		if (status != TRNG_NEW_RAND_AVAILABLE)
+			continue;
+
+		value = io_read32(xiphera_trng_base + RAND_REG);
+
+		/*
+		 * Ack that RNG value has been consumed and trigger new one to
+		 * be generated
+		 */
+		io_write32(xiphera_trng_base + CONTROL_REG, HOST_TO_TRNG_READ);
+		io_write32(xiphera_trng_base + CONTROL_REG,
+			   HOST_TO_TRNG_ENABLE);
+
+		break;
+	}
+
+	cpu_spin_unlock_xrestore(&trng_lock, exceptions);
+
+	return value;
+}
+
+/* This is a true RNG, no need for seeding */
+void plat_rng_init(void)
+{
+}
+
+TEE_Result crypto_rng_read(void *buf, size_t len)
+{
+	uint8_t *rngbuf = buf;
+	uint32_t val = 0;
+	size_t len_to_copy = 0;
+
+	assert(buf);
+	assert(xiphera_trng_base);
+
+	while (len) {
+		val = xiphera_trng_read32();
+		len_to_copy = MIN(len, sizeof(uint32_t));
+		memcpy(rngbuf, &val, len_to_copy);
+		rngbuf += len_to_copy;
+		len -= len_to_copy;
+	}
+
+	return TEE_SUCCESS;
+}
+
+uint8_t hw_get_random_byte(void)
+{
+	uint8_t data = 0;
+
+	assert(xiphera_trng_base);
+
+	data = xiphera_trng_read32() & 0xFF;
+
+	return data;
+}
+
+static TEE_Result xiphera_trng_probe(const void *fdt, int node,
+				     const void *compat_data __unused)
+{
+	int dt_status = _fdt_get_status(fdt, node);
+	uint32_t status = 0;
+	size_t size = 0;
+
+	/* Skip non-secure instances */
+	if (dt_status != DT_STATUS_OK_SEC)
+		return TEE_ERROR_GENERIC;
```

Comment:
Not this has side effect of:
```
E/TC:0 0 probe_driver_node:364 Failed to probe xiphera_trng on node xiphera_trng@a3f00000: 0xffff0000
I/TC: HUK ready
E/TC:0 0 probe_dt_drivers:701 Probe sequence result: 0
E/TC:0 0 print_probe_list:320 |- Driver xiphera_trng on node xiphera_trng@a3f00000 failed
I/TC: Primary CPU switching to normal world boot
```
We are using system device tree that defines more devices than what just OP-TEE needs.

We have dedicated TRNG instance for Linux and one for OP-TEE.

I used similar method for this return value as what was in existing driver but I would like to silence the error message as it is not really a error message. Any advice on this?

Reply:
It looks like we may need a `TEE_ERROR_NODE_DISABLED` error code.

Reply:
OK. Let's see how that unravels.

Reply:
Added commit for the feature. Please have a look.

Now output is silent on this matter.